versioning/worker-versions.md

```diff
@@ -0,0 +1,139 @@
+
+Temporal workflow authors experience a number of difficulties when it comes to
+changing workflow code. The current techniques for versioning workflows are:
+
+**GetVersion & patch APIs**
+These APIs allow branching within the same workflow code to handle the presence
+or absence of a certain change. See them for [TS](https://docs.temporal.io/docs/typescript/patching#typescript-sdk-patching-api), [Java](https://docs.temporal.io/docs/java/versioning/), and [Go](https://docs.temporal.io/docs/go/versioning/).
+
+These APIs are confusing and challenging to use correctly, and clutter up code.
+We advise users to generally use them only to fix bugs that would otherwise prevent
+the workflow from completing.
+
+**Versioned task queues**
+Conceptually simpler, but operationally more effort, is to version task queues.
+IE: A worker operating on task queue `foo-v1` is changed, and when redeployed
+points at `foo-v2`.
+
+This avoids complicating the workflow code itself, but means users need to be
+aware of when all workflows on `foo-v1` are complete, allowing them to decommission
+the old workers. It also means potentially deploying larger numbers of workers.
+
+It also requires updating clients who are starting workflows to point at the
+newer task queue name.
+
+**Versioned workflow names**
+This approach is much like versioning task queue names, but trades off some of
+the operational complexity for some code maintenance complexity.
+
+Rather than needing to keep old workers around, users now need to keep the
+old code around until no more workflows using the old type are open.
+
+They still have the same problem with needing to tell clients to use the new
+workflow name.
+
+## Improving the situation
+
+The `patch` type APIs can't be eliminated completely. There will always be 
+situations where a "broken" workflow needs to be fixed to make progress. Such
+fixes are typically less common than just making changes to workflow logic, though.
+
+To make the more common case easier, we will provide a set of features that
+make the "versioned task queues" approach much easier to use for changes to
+workflows which break history compatability (but not interface compatability).
+
+At a high level:
+* Provide a simple API for versioning entire workers (this proposal)
+* Allow clients to point at task queue `foo`, and automatically route new workflow executions
+  to the latest version of the `foo` queue.
+* Provide good notifications and information about when old workers may be
+  decommissioned
+* Ensure histories replayed on older versions fail fast with informative errors
+* Easy visibility into what versions deployed workers are using
+
+Building on this base set of features, we plan to eventually implement
+automatic safe-rollout of changes, and possibly tools like k8s operators
+which can automatically manage worker deployments.
+
+## SDK API changes
+
+The meat of this proposal is simple:
+**Add a version field to worker options in each SDK**
```

Comment:
_TLDR: I'd like to explore whether it is possible to deal with backwards-incompatibility of workflow code without relying on the human judgement element of semver-like versioning. In this comment I’m also making the assumption that it is possible to detect backwards-incompatibility via workflow replay._

My company runs an internal workflow automation platform built on Temporal. It's just vanilla Temporal with a few more opinionated layers on top like workflow interfaces defined in protobuf and a worker framework that handles authz, observability, deployment config, etc.

Among our customers, a common thread is that regardless of an engineer's seniority level or experience with Temporal, people sometimes forget to use workflow versioning, or may write a bug while attempting to do so. Recovering from a bad code change like this is tricky. So a while back we set out to solve the same issues described in this proposal, but with an added constraint: don't depend on human judgement to make worker deployments safe by default.

### Why not rely on a human specified version?

- The same engineers who forget to add version checks in workflow code can forget to update the worker's version.
- How and when to update the version requires perfect knowledge of both workflow code backwards compatibility rules AND the super-major/major/minor rules defined in this proposal. Failing to increment the version correctly when there has been a breaking change would appear to result in the same "recover from bad code change" complexity we set out to avoid.
- Adding, updating, or removing an interceptor can also break workflow replay. Changes to other library code can do the same. We try to avoid this stuff, but it's hard to ensure compatibility only through code review and workflow history snapshot/replay testing.
- Coordinating to increment a version among a large team of developers can be tricky, to the point that we've moved away from semver for releases of other internal services.

### Automated Breaking Change Detection

I think there are elements of this proposal and our internal approach that can be combined. First I'll describe our current approach to breaking change detection:

Each worker has two task queues, eg. `foo:canary` and `foo:main`. All production traffic flows to the `foo:main` queue. Our worker framework also automatically registers a "reflection" workflow on each task queue that allows querying a list of all workflows, activities, signals and queries supported by the worker, and a “replay” workflow that allows replaying any workflow ID.

During worker deployment, we first roll out the new version on the canary task queue. A generic “breaking change detection” workflow is then executed which:

1. Uses the reflection workflow to query a list of workflow and activity types registered on `foo:main`
2. Does the same to retrieve a list of workflow and activity types registered on `foo:canary`
3. Diffs the two responses to statically check for removed workflow/activity types etc.
4. Queries Temporal for a list of running workflow executions of each type found on `foo:main`, then replays each workflow's history on `foo:canary`.

If issues are surfaced by the breaking change detection workflow, then we block the deployment and issue a manual judgement. If a developer decides to roll out anyway, then we direct them to resources on batch reset or termination of broken workflows.

Note that we don’t view this as the end-all-be-all of developer experience. Most of these issues should be caught by continuous integration tests before a breaking change is ever merged.

### Alternative Proposal: Binary Checksums

There are some very interesting ideas here that have me rethinking the above approach. Some downsides of the canary/main replay testing technique are that:

- In the time we spend replaying workflow histories, more workflows may be started and others may make progress, resulting in a chance of breaking changes being missed.
- We may have thousands or more workflows to replay, in which case sampling may be necessary in order to reduce deployment latency. Again this increases our chances of missing breaking changes.
- We had to make large investments in tooling to make this work, and our solution may be incongruent with the practices of the larger Temporal community.

If this proposal were accepted as-is, then I think we might be able to simplify our internal solution by having our deployment tooling auto-increment workers by a major version for every change and skipping the replay parts of the breaking change detection workflow. This is driven by the assumption that any rollout has the possibility of containing a breaking change, and because we value the ability to recover gracefully over rollout speed (is this is the tradeoff between major and minor/patch version bumps)?

> Building on this base set of features, we plan to eventually implement automatic safe-rollout of changes

So I’m left wondering, is it possible to simplify things further for other users as well? For an alternative approach we could look to [worker binary checksums](https://github.com/temporalio/api/blob/336a14569b47d145300989c830cf50ab8ccad577/temporal/api/workflowservice/v1/request_response.proto#L212-L213), which are an existing method to track worker versions.

When breaking changes to workflow code are rolled out today, worker binary checksums are an important tool for recovering workflows. For example after a bad deployment, an operator can mark a binary as bad to prevent further progress being made at that version:

```bash
tctl namespace update --add_bad_binary sha256:deadbeef
```

 Then roll back the worker deployment and reset any workflows that had made progress at that version:

```bash
tctl workflow reset-batch --query “TaskQueue = ‘foo’ and ExecutionStatus = 'Running'" --only_non_deterministic --reset_type BadBinary --reason "bad worker version" --reset_bad_binary_checksum sha256:deadbeef
```

My thought is that we could combine binary checksums with a new “rollout strategy” worker option. Rollout strategy would default to the existing behavior of “any (non-bad) worker binary can poll for work”, but also include a “versioned cutover” option that matches the behavior described in this proposal for major version bumps. This has two key benefits:

- Keeps semantic versioning of workers optional and orthogonal to rollout strategy. Different rollout strategies may easily be applied in dev/staging/prod without changing versioning strategy between environments.
- Relies on an existing concept which is already key to more advanced workflow batch operations and recovery options. This means very little investment needed from anyone already using binary checksums, and leaves less room for confusion between use cases for semantic versions vs. binary versions.

Note that binary checksums are not sortable; Temporal server would probably need to keep track of the first observed poll time for the last _n_ checksums on a given task queue or within a given namespace so that we know which is the target “latest” version. An API for rollbacks would also be necessary so that an older checksum could be set to the target version. This could also be useful debugging info to surface in the web UI and for suggesting actions for workflow operators in the event of bad deploys.

I think this gets us pretty close to “automatic safe-rollout of changes”, since workflows would be automatically replayed before cutting over between every new worker version. Any non-deterministic code changes, missing workflow/activity/signal types, payload encoding issues, etc should raise errors during replay and thus keep those workflows polling on the previous worker version.

Reply:
Thanks for the very thoughtful feedback!

I think what you're getting at is totally reasonable, and the idea of making the cut-over work with just checksums so you can pick that over manual versioning if you prefer makes sense to me. The process you described of replaying existing histories to verify new versions are good is very much part of the automatic safe-rollout stuff we'll build later that I mentioned. So a lot of this is in-line with what we already are thinking of.

The meat of what you're getting at I think is really "what does safe rollout look like" which is the next step after this. I think the only alteration required to my proposal is to allow specifying compatibility more flexibly than just semver. For major changes, things are essentially exactly equivalent assuming you auto-bump major revisions (which, maybe that's the primary reason to _only_ use checksums).

So, if you wanted to do a safe-rollout of a non-major change using only checksums, maybe you have something like:

```typescript
Worker.create({ isCompatWith: ['deadbeef', 'feedbeef'], ...otherOptions })
```

When this worker starts up, it can
1. Tell server "Here's my checksum" so that server knows it is latest and begins queuing new workflow starts in that queue
2. Replays all (or some sample of, probably configurable) workflows on the queues it's compat with
3. If that works, tells server "passed compat check"
4. Begins polling (probably after polling a "have I passed yet?" check)

There are a few complications here, like you probably only want one worker to do all the replaying (or even more advanced, break up work among workers) so that other workers can immediately start processing new workflow starts, since we don't care about compat for these anyways. So you'd either have to explicitly nominate one worker as "leader" by passing in an env var or something, or possibly server just designates the first worker to connect as leader.

I think that could be really nice. Can you elaborate on "... "rollout strategy" worker option ..." and "Keeps semantic versioning of workers optional and orthogonal to rollout strategy. "? 

My feeling is that user-specified semver provides pretty much nothing except a pre-defined ordering of the versions which could be used to sanity check that checksums don't go "backward", and maybe just as a nice human-readable thing in the UI. 

I want to draw up some sequence diagrams of the above (I had done so for the semver-based thing but they aren't included in here) to think through it a bit more and we can use that do decide if we want to do this at all, or replace semver, or compliment semver.


I wrote the above after I wrote the below, but, replied to some specific things down there.

---

> If this proposal were accepted as-is, then I think we might be able to simplify our internal solution by having our deployment tooling auto-increment workers by a major version for every change

This is pretty much exactly what I imagined people who prefer your sort of approach would do.

Regarding:
> we value the ability to recover gracefully over rollout speed (is this is the tradeoff between major and minor/patch version bumps)?

Not even, really. In this proposal minor versions still get their own queue, though rather than only starting brand new workflows, new workflow _tasks_ may be started there. Really the only reason to use a minor version is because you have some bug that's causing workflows to get "stuck" and you want to unblock them, and for whatever reason you can't/don't want to reset instead. If pure deployment speed is your concern, you wouldn't even bump minor version here so you can keep using the same queue, but of course this also allows "old" workers to keep processing new tasks (which I assume generally people would prefer to avoid). You could bump patch purely for your own bookkeeping.

<br/>

One of the sticking points with the checksums is they're surprisingly difficult to reliably calculate for TypeScript and upcoming Python, but in practice it's probably solvable for everything but the most unusual setups.

The lack of a defined ordering is the other issue as you mentioned.
> keep track of the first observed poll time for the last n checksums on a given task queue or within a given namespace so that we know which is the target “latest” version.

I think that works the vast majority of the time, it's slightly error prone in terms of making sure you don't connect workers in some out-of-order fashion but that'd be a pretty odd situation to be in. No worse than some of the other things you could forget, and you need the rollback API anyway to undo that.

Reply:
I made a diagram of what his might look like
```mermaid
sequenceDiagram
actor Client
participant T as Temporal
participant W11 as v1.1 Worker

Note right of T: A v1 (0xbeef1) Worker is already established as current "latest"<br/> (not shown)
W11 ->> T: Upon connection: "I have bin id 0xbeef1.1 and I am compat with 0xbeef1" <br/> I use rollout strategy "wait for replay to pass"
T -> T: A new compat worker is connected, but has not been approved as compatible

Client ->> T: Start workflow on `foo`
T -> T: Latest for `foo` is 0xbeef11.1, so I will start the WF on that queue
Note right of T: Whether to queue starts & new tasks on the new ver or use the existing<br/> one can be a configurable part of rollout strat.

W11 ->> T: Poll for "Am I approved yet?" 
T -->> W11: No, and you are the leader

W11 ->> T: Fetch workflow histories from 0xbeef1 queue
T -->> W11: Histories
W11 -> W11: Replay histories

W11 ->> T: 0xdeadbeef1.1 has passed compatability check
W11 ->> T: Poll for "Am I approved yet?"  (all 1.1 workers will be doing this)
T -->> W11: Yes
Note left of W11: Now 1.1 workers know it is OK to start pulling real work

W11 ->> T: Poll on `foo-0xbeef1.1` queue
```